doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,222 @@
+# Errors without `errorKey`
+
+## Problem
+
+We often use a `Key * errorKey` argument to return error information via metadata (see related decisions).
+In `libelektra-core`, however, that would be inappropriate and sometimes even impossible (e.g., `keyNew`).
+
+Therefore, we need a different concept for errors, when an `errorKey` cannot be used.
+
+## Constraints
+
+- Using special return values as errors may be limited.
+  For example, only `NULL` is an invalid pointer and sometimes even that is a legitimate non-error result.
+- It should be easy to detect error cases and distinguish between errors, once an error is detected.
+- It should be hard to miss the fact that errors are possible.
+  Note, however, this does not mean checking for an error should be enforced.
+  There are many cases, where the caller knows an error is impossible, because of the previous code path.
+  The requirement is analogous to the `Key * errorKey` argument.
+  The presence of the argument makes it obvious that errors _can_ happen, but you can easily omit the error check when you want to.
+- Adding error indications should not make the API less convenient to use.
+
+## Assumptions
+
+- In many cases, it is possible to avoid error cases through the preceding code path.
+- When an error occurred, the kind error can be distinguished after the fact in many cases.
+- Callers normally don't care about the exact error, unless they can do something about it.
+  This could be:
+  - reporting the error to a user
+  - switching to a different code path
+  - (during debugging) changing the code
+- Functions can be written in a such a way that they detect and report all errors without permanent effects.
+
+## Considered Alternatives
+
+### Extra argument
+
+All functions that report errors take a `int * error` argument to which errors can be written.
+
+This makes the API much more convoluted to use.
+
+A benefit would be the return value remains fully usable.
+
+### Thread Local Storage
+
+Use a field in thread local storage (TLS) to store errors similar to `errno`.
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Pragmatic solution
```

Comment:
1. "next to memory errors" doesn't make it clear whether every function has 1 (for everything) or 2 (for memory and other) error values.
2. "Redesign API" is a new idea, which I'll add. But I believe it would be unnecessarily restrictive. For example, you couldn't have a `elektraCheckEscapedName` function that returns different values for different problems with the name (illegal escape, invalid namespace, etc.). You could argue those are not "error cases", but then what's the point of the restriction in the first place.
3. "Require users to check preconditions" goes partly in the direction of "Pragmatic Solution". I don't quite understand how it would work though. So the user has to do all the precondition checks. But what happens if they don't? the part about "against our principle" suggests, the APIs function would not check the preconditions. So do we just try and hope there isn't a segfault?
   If that's the suggestions, then sure I can add that. But that seems pretty stupid and I don't really think adding it as an option helps in anyway.
   If the API functions should still check the preconditions they obviously have to react somehow, which probably still means returning an error, but maybe just always the same error. In that case, it's basically "Pragmatic Solution".
   So I think I won't add this one...

Reply:
> For example, you couldn't have a elektraCheckEscapedName function that returns different values for different problems with the name (illegal escape, invalid namespace, etc.). 

Something like this should be written in the decision. (Btw. it was just by accident that I opened this hidden/resolved conversation.)